File: tests/test_7_2_1_transaction_graph.py
"""
Tests for Task 7.2.1: TinkerTransactionGraph — ACID Transaction Support.

Validates commit/rollback lifecycle, snapshot semantics, and context-manager
protocol directly against TinkerCat's transaction API.
See docs/project/changelog/task-7.2.1-transaction-graph.adoc.
"""

import pytest

from tinkercat import TinkerCat


@pytest.fixture
def g():
    graph = TinkerCat()
    yield graph
    graph.close()


def test_commit_persists_changes(g):
    tx = g.tx()
    tx.open()
    g.add_vertex("person", name="Alice")
    tx.commit()
    assert g.vertex_count == 1


def test_rollback_reverts_changes(g):
    tx = g.tx()
    tx.open()
    g.add_vertex("person", name="Alice")
    tx.rollback()
    assert g.vertex_count == 0


def test_double_open_raises(g):
    tx = g.tx()
    tx.open()
    with pytest.raises(Exception):
        tx.open()
    tx.rollback()


def test_commit_without_open_raises(g):
    tx = g.tx()
    with pytest.raises(Exception):
        tx.commit()


def test_rollback_without_open_raises(g):
    tx = g.tx()
    with pytest.raises(Exception):
        tx.rollback()


def test_close_on_open_transaction_rolls_back(g):
    tx = g.tx()
    tx.open()
    g.add_vertex("node")
    tx.close()
    assert g.vertex_count == 0


def test_tx_is_open_property(g):
    tx = g.tx()
    assert not tx.is_open
    tx.open()
    assert tx.is_open
    tx.commit()
    assert not tx.is_open


def test_context_manager_commits_on_success(g):
    with g.tx():
        g.add_vertex("person", name="Alice")
    assert g.vertex_count == 1


def test_context_manager_rolls_back_on_exception(g):
    with pytest.raises(ValueError):
        with g.tx():
            g.add_vertex("person")
            raise ValueError("deliberate error")
    assert g.vertex_count == 0


def test_rollback_restores_property_values(g):
    v = g.add_vertex("item", score=10)
    vid = v.id
    tx = g.tx()
    tx.open()
    v.set_property("score", 99)
    tx.rollback()
    restored = g.get_vertex(vid)
    assert restored.value("score") == 10


def test_committed_changes_survive_new_rollback(g):
    with g.tx():
        g.add_vertex("node")
    tx2 = g.tx()
    tx2.open()
    g.add_vertex("node")
    tx2.rollback()
    assert g.vertex_count == 1

File: tests/kotlin/test_7_2_1_transaction_graph.kt
package org.apache.tinkerpop.gremlin.tinkercat.tests

import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.types.shouldBeInstanceOf
import org.apache.tinkerpop.gremlin.tinkercat.structure.TinkerCat
import org.apache.tinkerpop.gremlin.tinkercat.structure.TinkerTransaction
import org.apache.tinkerpop.gremlin.tinkercat.structure.TinkerVertex

/**
 * Tests for Task 7.2.1: TinkerTransactionGraph — ACID Transaction Support.
 *
 * TinkerCat.tx() returns a TinkerTransaction stub whose every method throws
 * UnsupportedOperationException. Tests marked "!" are pending until a real
 * transaction implementation lands; their bodies compile because TinkerTransaction
 * is defined in source (not just in this test).
 * Mirrors tests/python/test_7_2_1_transaction_graph.py.
 */
class Test_7_2_1_TransactionGraph : StringSpec({

    lateinit var graph: TinkerCat

    beforeTest { graph = TinkerCat.open() }

    "tx() returns a TinkerTransaction stub" {
        val tx = graph.tx()
        tx.shouldBeInstanceOf<TinkerTransaction>()
    }

    "tx().open() raises UnsupportedOperationException — not yet implemented" {
        val tx = graph.tx()
        val caught = runCatching { tx.open() }.exceptionOrNull()
        (caught is UnsupportedOperationException) shouldBe true
    }

    "!commit persists changes" {
        val tx = graph.tx()
        tx.open()
        graph.addVertex(TinkerVertex.T.label, "person", "name", "Alice")
        tx.commit()
        graph.vertices.size shouldBe 1
    }

    "!rollback reverts changes" {
        val tx = graph.tx()
        tx.open()
        graph.addVertex(TinkerVertex.T.label, "person", "name", "Alice")
        tx.rollback()
        graph.vertices.size shouldBe 0
    }

    "!double open raises" {
        val tx = graph.tx()
        tx.open()
        runCatching { tx.open() }.isFailure shouldBe true
        tx.rollback()
    }

    "!commit without open raises" {
        val tx = graph.tx()
        runCatching { tx.commit() }.isFailure shouldBe true
    }

    "!rollback without open raises" {
        val tx = graph.tx()
        runCatching { tx.rollback() }.isFailure shouldBe true
    }

    "!context manager commits on success" {
        graph.tx().use { tx ->
            tx.open()
            graph.addVertex(TinkerVertex.T.label, "person", "name", "Alice")
        }
        graph.vertices.size shouldBe 1
    }

    "!context manager rolls back on exception" {
        runCatching {
            graph.tx().use { tx ->
                tx.open()
                graph.addVertex(TinkerVertex.T.label, "person")
                throw RuntimeException("deliberate error")
            }
        }
        graph.vertices.size shouldBe 0
    }

    "!rollback restores property values" {
        val v = graph.addVertex(TinkerVertex.T.label, "item", "score", 10) as TinkerVertex
        val tx = graph.tx()
        tx.open()
        v.property("score", 99)
        tx.rollback()
        v.value<Int>("score") shouldBe 10
    }

    "!committed changes survive new rollback" {
        graph.tx().use { tx ->
            tx.open()
            graph.addVertex(TinkerVertex.T.label, "node")
        }
        val tx2 = graph.tx()
        tx2.open()
        graph.addVertex(TinkerVertex.T.label, "node")
        tx2.rollback()
        graph.vertices.size shouldBe 1
    }
})
